extension/InspectionDoc/src/Infrastructure/Repository/InspectionDocRepository.php

```diff
@@ -0,0 +1,141 @@
+
+use InspectionDoc\Entity\InspectionDoc;
+use InspectionDoc\Entity\InspectionDocInterface;
+use InspectionDoc\Infrastructure\DataProvider\InspectionDocDataProviderInterface;
+
+class InspectionDocRepository implements InspectionDocRepositoryInterface
+{
+    /**
+     * @return array<string, array<mixed>>
+     */
+    protected ?array $indexedInspectionDocs = null;
+
+    /**
+     * @var array<string, \InspectionDoc\Entity\InspectionDocInterface>
+     */
+    protected array $inspectionDocsCache = [];
+
+    /**
+     * @var \InspectionDoc\Infrastructure\DataProvider\InspectionDocDataProviderInterface
+     */
+    protected InspectionDocDataProviderInterface $inspectionDocDataProvider;
+
+    /**
+     * @param \InspectionDoc\Infrastructure\DataProvider\InspectionDocDataProviderInterface $inspectionDocDataProvider
+     */
+    public function __construct(InspectionDocDataProviderInterface $inspectionDocDataProvider)
+    {
+        $this->inspectionDocDataProvider = $inspectionDocDataProvider;
+    }
+
+    /**
+     * @param string $errorCode
+     *
+     * @return \InspectionDoc\Entity\InspectionDocInterface|null
+     */
+    public function findByErrorCode(string $errorCode): ?InspectionDocInterface
+    {
+        $indexedInspectionDocs = $this->getIndexedInspectionDocs();
+
+        $inspectionId = $this->findInspectionIdByErrorCode(array_keys($indexedInspectionDocs), $errorCode);
+
+        if ($inspectionId === null) {
+            return null;
+        }
+
+        $inspectionDoc = $indexedInspectionDocs[$inspectionId];
+
+        return $this->createInspectionDoc($inspectionId, $inspectionDoc);
+    }
+
+    /**
+     * @param array $inspectionsIds
+     * @param string $errorCode
+     *
+     * @return string|null
+     */
+    protected function findInspectionIdByErrorCode(array $inspectionsIds, string $errorCode): ?string
+    {
+        $inspectionIdCandidates = [];
+
+        foreach ($inspectionsIds as $inspectionsId) {
+            if (!$this->strStartsWith($errorCode, $inspectionsId)) {
+                continue;
+            }
+
+            $inspectionIdCandidates[] = $inspectionsId;
+        }
+
+        return $this->getMaxLengthMatch($inspectionIdCandidates);
+    }
+
+    /**
+     * @param string $haystack
+     * @param string $needle
+     *
+     * @return bool
+     */
+    protected function strStartsWith(string $haystack, string $needle): bool
+    {
+        return strncmp($haystack, $needle, strlen($needle)) === 0;
+    }
+
+    /**
+     * @param array<string> $inspectionsIds
+     *
+     * @return string|null
+     */
+    protected function getMaxLengthMatch(array $inspectionsIds): ?string
+    {
+        usort($inspectionsIds, static fn (string $a, string $b): int => -1 * (strlen($a) <=> strlen($b)));
+
+        return $inspectionsIds[0] ?? null;
+    }
+
+    /**
+     * @return array<string, array<mixed>>
+     */
+    protected function getIndexedInspectionDocs(): array
```

Comment:
As it has the potential to have a bug because of the state. Also, it breaks SRP principle violation, I'd suggest extracting it to the `IndexedInspectionDocsStorage` class and cover with unit tests.

Reply:
It's just a cache for performance. Nothing happen when we create another instance.

Reply:
Agree, but it should be part of `JsonFileInspectionDocDataLoader`. The loader is responsible for the current data structure.